File: elixir/list-ops/list_ops.exs
defmodule ListOps do
  # Please don't use any external modules (especially List) in your
  # implementation. The point of this exercise is to create these basic functions
  # yourself.
  #
  # Note that `++` is a function from an external module (Kernel, which is
  # automatically imported) and so shouldn't be used either.

  @spec count(list) :: non_neg_integer
  def count([]), do: 0
  def count([_|t]), do: 1 + count(t)

  @spec reverse(list) :: list
  def reverse(list, reversed \\ [])
  def reverse([], reversed), do: reversed
  def reverse([h|t], reversed), do: reverse(t, [h|reversed])

  @spec map(list, (any -> any)) :: list
  def map([], _), do: []
  def map([h|t], f), do: [f.(h) | map(t, f)]

  @spec filter(list, (any -> as_boolean(term))) :: list
  def filter([], _), do: []
  def filter([h|t], f) do
    if f.(h) do
      [h|filter(t, f)]
    else
      filter(t, f)
    end
  end

  @type acc :: any
  @spec reduce(list, acc, ((any, acc) -> acc)) :: acc
  def reduce([], acc, _), do: acc
  def reduce([h|t], acc, f), do: reduce(t, f.(h, acc), f)

  @spec append(list, list) :: list
  def append([], b), do: b
  def append([ah|at], b), do: [ah|append(at, b)]

  @spec concat([[any]]) :: [any]
  def concat([]), do: []
  def concat([h|t]), do: append(h, concat(t))
end

File: python/list-ops/list_ops.py
# Please, do not use the built-in python functions like map, reduce, len, etc.
# that solve the same problems and try to solve it yourself instead.


def map_clone(function, xs):
    return [function(x) for x in xs]


def length(xs):
    count = 0
    for x in xs:
        count += 1
    return count


def filter_clone(function, xs):
    return [x for x in xs if function(x)]


def reverse(xs):
    l = []
    if not xs:
        return l

    for x in xs:
        l.insert(0, x)

    if isinstance(xs, list):
        return l
    elif isinstance(xs, tuple):
        return tuple(l)


def append(xs, y):
    return xs + [y]


def foldl(function, xs, acc):
    for x in xs:
        acc = function(acc, x)

    return acc


def foldr(function, xs, acc):
    for x in reverse(xs):
        acc = function(x, acc)

    return acc


def flat(xs):
    for i, x in enumerate(xs):
        if isinstance(x, list):
            return flat(xs[:i] + x + xs[i+1:])

    return xs

def concat(xs, ys):
    if not isinstance(xs, list):
        xs = []
    if not isinstance(ys, list):
        ys = []

    return xs + ys
